api/v1/submissions/PKPSubmissionHandler.inc.php

```diff
@@ -901,4 +941,281 @@ public function deletePublication($slimRequest, $response, $args)
+    {
+        $request = $this->getRequest();
+        $submission = $this->getAuthorizedContextObject(Application::ASSOC_TYPE_SUBMISSION);
+
+        $publication = Repo::publication()->get((int) $args['publicationId']);
+        $author = Repo::author()->get((int) $args['contributorId']);
+
+        if (!$publication) {
+            return $response->withStatus(404)->withJsonError('api.404.resourceNotFound');
+        }
+
+        if (!$author) {
+            return $response->withStatus(404)->withJsonError('api.404.resourceNotFound');
+        }
+
+        if ($submission->getId() !== $publication->getData('submissionId')) {
+            return $response->withStatus(403)->withJsonError('api.publications.403.submissionsDidNotMatch');
+        }
+
+        if ($publication->getId() !== $author->getData('publicationId')) {
+            return $response->withStatus(404)->withJsonError('api.404.resourceNotFound');
+        }
+
+        return $response->withJson(
+            Repo::author()->getSchemaMap()->map($author),
+            200
+        );
+    }
+
+    /**
+     * Get all publication's contributors
+     *
+     * @param $slimRequest Request Slim request object
+     * @param $response Response object
+     * @param $args array arguments
+     *
+     * @return Response
+     */
+    public function getContributors($slimRequest, $response, $args)
+    {
+        $request = $this->getRequest();
+
+        $submission = $this->getAuthorizedContextObject(Application::ASSOC_TYPE_SUBMISSION);
+        $publication = Repo::publication()->get((int) $args['publicationId']);
+
+        if (!$publication) {
+            return $response->withStatus(404)->withJsonError('api.404.resourceNotFound');
+        }
+
+        if ($submission->getId() !== $publication->getData('submissionId')) {
+            return $response->withStatus(403)->withJsonError('api.publications.403.submissionsDidNotMatch');
+        }
+
+        $collector = Repo::author()->getCollector();
+        $collector->filterByPublicationIds([$publication->getId()]);
+        $authors = Repo::author()->getMany($collector);
+
+        return $response->withJson([
+            'itemsMax' => Repo::author()->getCount($collector->limit(null)->offset(null)),
+            'items' => Repo::author()->getSchemaMap()->summarizeMany($authors),
+        ], 200);
+    }
+
+    /**
+     * Add a new contributor to publication
+     *
+     * This will create a new contributor from scratch.
+     *
+     * @param $slimRequest Request Slim request object
+     * @param $response Response object
+     * @param array $args arguments
+     *
+     * @return Response
+     */
+    public function addContributor($slimRequest, $response, $args)
+    {
+        $request = $this->getRequest();
+        $submission = $this->getAuthorizedContextObject(Application::ASSOC_TYPE_SUBMISSION);
+        $publication = Repo::publication()->get((int) $args['publicationId']);
```

Comment:
I think you'll need to check that the publication exists and is assigned to the right submission here.

Reply:
Added 

```
if (!$publication) {
  return $response->withStatus(404)->withJsonError('api.404.resourceNotFound');
}

if ($submission->getId() !== $publication->getData('submissionId')) {
  return $response->withStatus(403)->withJsonError('api.publications.403.submissionsDidNotMatch');
}
```